api/lib/domain/usecases/generate-organization-learners-username-and-temporary-password.js

```diff
@@ -0,0 +1,165 @@
+import { UserNotAuthorizedToUpdatePasswordError } from '../../../src/shared/domain/errors.js';
+import { OrganizationLearnerPasswordResetDTO } from '../../../src/shared/domain/models/OrganizationLearnerPasswordResetDTO.js';
+import {
+  ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
+  ORGANIZATION_LEARNER_WITHOUT_USERNAME_CODE,
+} from '../constants/generate-organization-learners-username-and-temporary-password-errors.js';
+
+const generateOrganizationLearnersUsernameAndTemporaryPassword = async function ({
+  organizationId,
+  organizationLearnersId,
+  userId,
+  cryptoService,
+  passwordGenerator,
+  userReconciliationService,
+  authenticationMethodRepository,
+  organizationRepository,
+  organizationLearnerRepository,
+  userRepository,
+}) {
+  const errorMessage = `User ${userId} cannot reset passwords of some students in organization ${organizationId}`;
+  const organization = await organizationRepository.get(organizationId);
+  const organizationLearners = await organizationLearnerRepository.findByIds({ ids: organizationLearnersId });
+  const userIds = organizationLearners.map((organizationLearner) => organizationLearner.userId);
+  const users = await userRepository.getByIds(userIds);
+  let organizationLearnerIdentities = _buildOrganizationLearnerIdentities({ organizationLearners, users });
+
+  _assertEachOrganizationLearnersBelongToOrganization({
+    errorMessage,
+    organizationId,
+    organizationLearners,
+  });
+
+  if (organization.isGarIdentityProvider) {
+    organizationLearnerIdentities = await _generateAndUpdateUsernameForOrganizationLearnerIdentities({
+      organizationLearnerIdentities,
+      userReconciliationService,
+      userRepository,
+    });
+  }
+
+  const userIdWithPasswords = await _generateAndUpdateUsersWithTemporaryPassword({
+    errorMessage,
+    organizationLearnerIdentities,
+    authenticationMethodRepository,
+    cryptoService,
+    passwordGenerator,
+  });
+
+  return _buildOrganizationLearnerPasswordResetDTOs({ organizationLearnerIdentities, userIdWithPasswords });
+};
+
+function _assertEachOrganizationLearnersBelongToOrganization({ errorMessage, organizationId, organizationLearners }) {
+  const organizationLearnersBelongsToOrganization = organizationLearners.every(
+    (organizationLearner) => organizationLearner.organizationId === organizationId,
+  );
+
+  if (!organizationLearnersBelongsToOrganization) {
+    throw new UserNotAuthorizedToUpdatePasswordError(
+      errorMessage,
+      ORGANIZATION_LEARNER_DOES_NOT_BELONG_TO_ORGANIZATION_CODE,
+    );
+  }
+}
+
+function _buildOrganizationLearnerIdentities({ organizationLearners, users }) {
+  const userIdUserMap = new Map(users.map((user) => [user.id, user]));
+  return organizationLearners.map(({ division, firstName, lastName, birthdate, userId }) => {
+    const user = userIdUserMap.get(userId);
```

Comment:
Au lieu de passer par une `Map` transitoire, on peut utiliser `Array#find`
```suggestion
    const user = users.find(user => user.id === userId);
```

Reply:
Fait sur ce commit : https://github.com/1024pix/pix/pull/10096/commits/645c501347f1ccaed6ce787bbc2432682967788f